activation/atxdb.go

```diff
@@ -366,111 +357,88 @@ func (db *ActivationDb) ActiveSetSize(epochId types.EpochId) (uint32, error) {
-	ids, err := db.atxs.Get(epochId.ToBytes())
-	var atxs []types.AtxId
+func (db *ActivationDb) updateTopAtxIfNeeded(atx *types.ActivationTx) error {
+	currentTopAtx, err := db.getTopAtx()
+	if err != nil && err != database.ErrNotFound {
+		return fmt.Errorf("failed to get current ATX: %v", err)
+	}
+	if err == nil && currentTopAtx.LayerId >= atx.PubLayerIdx {
+		return nil
+	}
+
+	newTopAtx := atxIdAndLayer{
+		AtxId:   atx.Id(),
+		LayerId: atx.PubLayerIdx,
+	}
+	topAtxBytes, err := types.InterfaceToBytes(&newTopAtx)
 	if err != nil {
-		//epoch doesnt exist, need to insert new layer
-		ids = []byte{}
-		atxs = make([]types.AtxId, 0, 0)
-	} else {
-		atxs, err = decodeAtxIds(ids)
-		if err != nil {
-			return errors.New("could not get all atxs from database ")
-		}
+		return fmt.Errorf("failed to marshal top ATX: %v", err)
 	}
-	atxs = append(atxs, atx)
-	w, err := encodeAtxIds(atxs)
+
+	err = db.atxs.Put([]byte(topAtxKey), topAtxBytes)
 	if err != nil {
-		return errors.New("could not encode layer atx ids")
+		return fmt.Errorf("failed to store top ATX: %v", err)
 	}
-	return db.atxs.Put(epochId.ToBytes(), w)
+	return nil
+}
+
+func (db ActivationDb) getTopAtx() (atxIdAndLayer, error) {
+	topAtxBytes, err := db.atxs.Get([]byte(topAtxKey))
+	if err != nil {
+		return atxIdAndLayer{}, err
+	}
+	var topAtx atxIdAndLayer
+	err = types.BytesToInterface(topAtxBytes, &topAtx)
+	if err != nil {
+		return atxIdAndLayer{}, fmt.Errorf("failed to unmarshal top ATX: %v", err)
+	}
+	return topAtx, nil
 }
 
 func getNodeIdKey(id types.NodeId) []byte {
 	return []byte(id.Key)
 }
 
-// addAtxToNodeIdSorted inserts activation atx id by node it is not thread safe and needs to be called under a global lock
-func (db *ActivationDb) addAtxToNodeIdSorted(nodeId types.NodeId, atx *types.ActivationTx) error {
+// addAtxToNodeId inserts activation atx id by node
+func (db *ActivationDb) addAtxToNodeId(nodeId types.NodeId, atx *types.ActivationTx) error {
 	key := getNodeIdKey(nodeId)
-	ids, err := db.atxs.Get(key)
-	var atxs []types.AtxId
-	if err != nil {
-		//layer doesnt exist, need to insert new layer
-		ids = []byte{}
-		atxs = make([]types.AtxId, 0, 0)
-	} else {
-		atxs, err = decodeAtxIds(ids)
-		if err != nil {
-			return errors.New("could not get all atxs from database ")
-		}
-	}
-
-	// append needs to be sorted
-	atxs = append(atxs, atx.Id())
-	l := len(atxs)
-	if l > 1 {
-		lastAtx, err := db.getAtxUnlocked(atxs[l-2])
-		if err != nil {
-			return errors.New("could not get atx from database ")
-		}
-		if atx.Sequence < lastAtx.Sequence {
-			sort.Slice(atxs, func(i, j int) bool {
-				atx1, err := db.getAtxUnlocked(atxs[i])
-				if err != nil {
-					panic("inconsistent db")
-				}
-				atx2, err := db.getAtxUnlocked(atxs[j])
-				if err != nil {
-					panic("inconsistent db")
-				}
-
-				return atx1.Sequence < atx2.Sequence
-			})
-		}
-	}
-
-	w, err := encodeAtxIds(atxs)
+	err := db.atxs.Put(key, atx.Id().Bytes())
```

Comment:
do you need to validate that it is indeed the latest atx that this node published?

Reply:
Contextual validation ensures that we store ATXs in the correct order.

It may have published another ATX, but this is the latest one in the DB.